WalletWasabi/Services/WalletService.cs

```diff
@@ -315,29 +322,27 @@ public HdPubKey GetReceiveKey(string label, IEnumerable<HdPubKey> dontTouch = nu
 			return ret;
 		}
 
-		public List<SmartCoin> GetHistory(SmartCoin coin, IEnumerable<SmartCoin> current)
+		public List<SmartCoin> GetHistory(SmartCoin coin, List<SmartCoin> current, ILookup<Script, SmartCoin> lookupScriptPubKey, ILookup<uint256, SmartCoin> lookupSpenderTransactionId, ILookup<uint256, SmartCoin> lookupTransactionId)
 		{
 			Guard.NotNull(nameof(coin), coin);
 			if (current.Contains(coin))
 			{
-				return current.ToList();
+				return current;
 			}
+
 			var history = current.Concat(new List<SmartCoin> { coin }).ToList(); // Fhe coin is the first elem in its history.
 
 			// If the script is the same then we have a match, no matter of the anonimity set.
-			foreach (var c in Coins)
+			foreach (var c in lookupScriptPubKey[coin.ScriptPubKey])
 			{
-				if (c.ScriptPubKey == coin.ScriptPubKey)
+				if (!history.Contains(c))
```

Comment:
Given `history` is a `List<SmartCoin>` it requires walk all the list in order to know if the list does not contain a coin, this is O(n) (take into account this is in a `for loop`. I think it could worth to consider using a `hash table` instead O(1).

Reply:
Did you read the conversation on this issue? https://github.com/zkSNACKs/WalletWasabi/pull/1076#issuecomment-453973584

Reply:
you got me! No, I didnt. Anyway, if we don't want to lose the order I think we could keep the order with each coin, does it make sense?

Reply:
To my understanding of the above conversation, David changed back to List, because with the lookup it didn't make sense to keep around performance wise, which means there's nothing to optimize anymore?

Reply:
I am sure there is a better solution. But according to long term goals for now it is good this way. We would only save a few seconds(<2 sec) or less on a low prio task.